README.md
## WalletConnect V1

Implementation of [WalletConnect protocol V1](https://docs.walletconnect.com/tech-spec) in Kotlin.

- Heavily uses Kotlin **Coroutines**.
- **Extendable**, you can provide your own implementation of almost anything
- Can be used in any **Kotlin** or **Android** project. Only Android [sample app](sample) is provided.
- ⚠️ Warning: Usage from Java projects is not tested
- ⚠️ Warning: APIs are not final yet, breaking changes should be expected

### Install

[![Release](https://jitpack.io/v/jemshit/WalletConnect.svg)](https://jitpack.io/#jemshit/WalletConnect)

**1. DApp & Wallet**

```kotlin
implementation("com.jemshit.walletconnect:walletconnect:x.y.z")
```

You can provide your own implementation of [DApp](walletconnect-core/src/main/java/walletconnect/core/DApp.kt)
and [Wallet](walletconnect-core/src/main/java/walletconnect/core/Wallet.kt)

**2. Socket**

```kotlin
implementation("com.jemshit.walletconnect:walletconnect-socket-scarlet:x.y.z")
```

You can provide your own implementation
of [Socket](walletconnect-core/src/main/java/walletconnect/core/socket/Socket.kt)

**3. Adapter**

```kotlin
implementation("com.jemshit.walletconnect:walletconnect-adapter-gson:x.y.z")
// or
implementation("com.jemshit.walletconnect:walletconnect-adapter-moshi:x.y.z")
```

You can provide your own implementation
of [JsonAdapter](walletconnect-core/src/main/java/walletconnect/core/adapter/JsonAdapter.kt)

**4. Session Store**

```kotlin
implementation("com.jemshit.walletconnect:walletconnect-store-file:x.y.z")
// or (Android only)
implementation("com.jemshit.walletconnect:walletconnect-store-prefs:x.y.z")
```

You can provide your own implementation
of [SessionStore](walletconnect-core/src/main/java/walletconnect/core/session_state/SessionStore.kt)

**5. Custom Request Models**

```kotlin
implementation("com.jemshit.walletconnect:walletconnect-requests:x.y.z")
```

**6. Other**

- Provide [Logger](walletconnect-core/src/main/java/walletconnect/core/util/Logger.kt)
    - Default
      provided: [ConsoleLogger](walletconnect-core/src/main/java/walletconnect/core/util/logger_impl/ConsoleLogger.kt)
    - Default
      provided: [EmptyLogger](walletconnect-core/src/main/java/walletconnect/core/util/logger_impl/EmptyLogger.kt)
    - Android sample: [AndroidLogger](sample/src/main/java/walletconnect/sample/impl/AndroidLogger.kt)

- Provide [DispatcherProvider](walletconnect-core/src/main/java/walletconnect/core/util/DispatcherProvider.kt)
    - Android
      sample: [AndroidDispatcherProvider](sample/src/main/java/walletconnect/sample/impl/AndroidDispatcherProvider.kt)

**7. Base module to provide your own implementations**

```kotlin
implementation("com.jemshit.walletconnect:walletconnect-core:x.y.z")
```

### Usage

Make sure to **check Documentation** of Functions & Models. They contain helpful information.

<details>
<summary>Create DApp/Wallet</summary>

```kotlin
fun createDApp(sessionStoreName: String)
        : DApp {
    return DAppManager(
            socket = createSocket(),
            sessionStore = createSessionStore(sessionStoreName),
            jsonAdapter = createJsonAdapter(),
            dispatcherProvider,
            logger
    )
}

fun createWallet(sessionStoreName: String)
        : Wallet {
    return WalletManager(
            socket = createSocket(),
            sessionStore = createSessionStore(sessionStoreName),
            jsonAdapter = createJsonAdapter(),
            dispatcherProvider,
            logger
    )
}

fun createSocketService(url: String,
                        lifecycleRegistry: LifecycleRegistry)
        : SocketService {

    // you can ignore this
    val interceptor = HttpLoggingInterceptor().apply {
        level = HttpLoggingInterceptor.Level.HEADERS
    }

    // change depending on your needs
    val okHttpClient = OkHttpClient.Builder()
            .callTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            // "https://bridge.walletconnect.org" -> i think BridgeServer responds with "missing or invalid socket data"
            // "https://safe-walletconnect.gnosis.io" -> ping works fine
            .pingInterval(4, TimeUnit.SECONDS)
            .addNetworkInterceptor(interceptor)
            .build()

    val webSocketFactory = okHttpClient.newWebSocketFactory(url)

    // you can use something else instead of Gson, make sure to provide SocketMessageTypeAdapter() & JsonRpcMethodTypeAdapter()
    val gson = GsonBuilder()
            .registerTypeAdapter(SocketMessageType::class.java, SocketMessageTypeAdapter())
            .registerTypeAdapter(JsonRpcMethod::class.java, JsonRpcMethodTypeAdapter())
            .create()

    val scarlet = Scarlet.Builder()
            .webSocketFactory(webSocketFactory)
            .addMessageAdapterFactory(GsonMessageAdapter.Factory(gson))
            .addStreamAdapterFactory(FlowStreamAdapter.Factory)
            .backoffStrategy(ExponentialBackoffStrategy(initialDurationMillis = 1_000L,
                                                        maxDurationMillis = 8_000L))
            .lifecycle(lifecycleRegistry)
            .build()

    return scarlet.create(SocketService::class.java)
}

fun createSocket()
        : Socket {
    val gson = GsonBuilder()
            .registerTypeAdapter(SocketMessageType::class.java, SocketMessageTypeAdapter())
            .registerTypeAdapter(JsonRpcMethod::class.java, JsonRpcMethodTypeAdapter())
            .create()

    return SocketManager(
            socketServiceFactory = { url, lifecycleRegistry -> createSocketService(url, lifecycleRegistry) },
            gson,
            dispatcherProvider,
            logger
    )
}

fun createSessionStore(name: String)
        : SessionStore {

    // return anything that implements SessionStore

    //val sharedPrefs = requireContext().applicationContext.getSharedPreferences(name, Context.MODE_PRIVATE)
    //return SharedPrefsSessionStore(
    //        sharedPrefs,
    //        dispatcherProvider,
    //        logger
    //)

    return FileSessionStore(
            File(requireContext().filesDir, "$name.json"),
            dispatcherProvider,
            logger
    )
}

fun createJsonAdapter()
        : JsonAdapter {

    // return anything that implements JsonAdapter. Make sure to provide SocketMessageTypeAdapter() & JsonRpcMethodTypeAdapter()

    //val gson = GsonBuilder()
    //        .registerTypeAdapter(SocketMessageType::class.java, SocketMessageTypeAdapter())
    //        .registerTypeAdapter(JsonRpcMethod::class.java, JsonRpcMethodTypeAdapter())
    //        .serializeNulls()
    //        .create()
    //
    //return GsonAdapter(gson)


    val moshi = Moshi.Builder()
            .add(walletconnect.adapter.moshi.type_adapter.SocketMessageTypeAdapter())
            .add(walletconnect.adapter.moshi.type_adapter.JsonRpcMethodTypeAdapter())
            .addLast(KotlinJsonAdapterFactory())
            .build()
    return MoshiAdapter(moshi)
}
```

</details>

<details>
<summary>DApp</summary>

**Init**

```kotlin
val connectionParams = ConnectionParams(
        topic = UUID.randomUUID().toString(), // unique topic = unique session
        version = "1",
        // "https://bridge.walletconnect.org" -> when one peer deletes session while other peer is disconnected, 
        //     other peer never gets that message even after connecting. Also pings in socket is not supported
        bridgeUrl = "https://safe-walletconnect.gnosis.io",
        symmetricKey = "..." // 32 byte (64 char) encryption/decryption key. You can use `Cryptography.generateSymmetricKey().toHex()`
)

val initialSessionState = InitialSessionState(
        connectionParams,
        myPeerId = UUID.randomUUID().toString(),
        myPeerMeta = PeerMeta(
                name = "DApp",
                url = "https://dapp.com",
                description = "DApp Description",
                icons = listOf("https://www.dapp.com/img/Icon_Logotype_2.png")
        )
)

val dApp: DApp = createDApp(sessionStoreName = "...")
```

**Open Socket**

```kotlin
dApp.openSocketAsync(initialSessionState,
                     callback = ::onSessionCallback,
                     onOpen = { freshOpened ->
                         // sendSessionRequest() etc...
                     })
// or
coroutineScope.launch(dispatcherProvider.io()) {
    val freshOpened = dApp.openSocket(initialSessionState,
                                      callback = ::onSessionCallback)
    // sendSessionRequest() etc...
}
```

**Close**

```kotlin
dApp.closeAsync(deleteLocal = false,
                deleteRemote = false,
                onClosed = { freshClosed ->
                    // ..
                })
// or 
coroutineScope.launch(dispatcherProvider.io()) {
    val freshClosed = dApp.close(deleteLocal = false,
                                 deleteRemote = false)
    // ...
}

// close & delete session
dApp.closeAsync(deleteLocal = true,
                deleteRemote = true,
                onClosed = { freshClosed ->
                    // ..
                })
```

**Session Request**

```kotlin
dApp.sendSessionRequest(chainId)
```

**Sign Request**

```kotlin
coroutineScope.launch(dispatcherProvider.io()) {
    val ethSign = EthSign(address = "...",
                          message = "...", // raw string for SignType.Sign, hex string for SignType.PersonalSign
                          type = SignType.Sign) // or SignType.PersonalSign

    // ethSign.validate()

    val messageId: Long? = dApp.sendRequest(
            method = ethSign.type.toMethod(),
            data = ethSign.toList(),
            itemType = String::class.java
    )
    // You can store Map<messageId, MyCallback>, so when you get response for this messageId in 'onSessionCallback', 
    //  you can invoke corresponding MyCallback
}

// or
dApp.sendRequestAsync(
        method = ethSign.type.toMethod(),
        data = ethSign.toList(),
        itemType = String::class.java,
        onRequested = {},
        onRequestError = {},
        onCallback = {}
)
```

**EthSendTransaction Request**

```kotlin
// Check sample for sending custom token using SmartContract address. 
// There is also gas estimation API example for Binance Smart Chain
// Check HexByteExtensions.kt for 'toHex' extension on String/Long/Int
fun createTransaction()
        : EthTransaction {
    return EthTransaction(
            from = approvedAddress,
            to = "0x621261D26847B423Df639848Fb53530025a008e8",
            data = "",
            chainId = approvedChainId.toHex(),

            gas = null,
            gasPrice = null,
            gasLimit = null,
            maxFeePerGas = null,
            maxPriorityFeePerGas = null,

            value = "0x" + BigInteger("10000000000000000").toString(16), // 1_000_000_000_000_000_000L.toHex(),
            nonce = null
    )
}

coroutineScope.launch(dispatcherProvider.io()) {
    // you can call EthTransaction.validate() before sending
    val messageId: Long? = dApp.sendRequest(
            EthRpcMethod.SendTransaction,
            data = listOf(createTransaction()),
            itemType = EthTransaction::class.java
    )
    // You can store Map<messageId, MyCallback>, so when you get response for this messageId in 'onSessionCallback', 
    //  you can invoke corresponding MyCallback
}
// or
dApp.sendRequestAsync(
        EthRpcMethod.SendTransaction,
        data = listOf(createTransaction()),
        itemType = EthTransaction::class.java,
        onRequested = {},
        onRequestError = {},
        onCallback = {}
)
```

**Custom Request**

```kotlin
// check JsonRpcMethod file for list of default provided
coroutineScope.launch(dispatcherProvider.io()) {
    val messageId: Long? = dApp.sendRequest(
            CustomRpcMethod("some_method_name"),
            data = listOf(MyClass()),
            itemType = MyClass::class.java
    )
}

// or
dApp.sendRequestAsync(
        CustomRpcMethod("some_method_name"),
        data = listOf(MyClass()),
        itemType = MyClass::class.java,
        onRequested = {},
        onRequestError = {},
        onCallback = {}
)
```

**Custom SocketMessage**

```kotlin
// 'dApp.sendRequest' uses 'encryptPayloadAndPublish' under the hood, you can use 'encryptPayloadAndPublish' directly
```

**Other**

```kotlin
// dApp.generateMessageId()
// dApp.getInitialSessionState()   // inherited from SessionLifecycle interface
// dApp.disconnectSocket()         // inherited from SessionLifecycle interface
// dApp.reconnectSocket()          // inherited from SessionLifecycle interface

// Callbacks
fun onSessionCallback(callbackData: CallbackData) {
    coroutineScope.launch(dispatcherProvider.ui()) {
        when (callbackData) {
            is SessionCallback -> {
                when (callbackData) {
                    // ...        
                }
            }
            is SocketCallback -> {
                when (callbackData) {
                    // ...
                }
            }
            is RequestCallback -> {
                when (callbackData) {
                    // ...
                }
            }
            is FailureCallback -> {
                // ...
            }
        }
    }
}

// DeepLink to Wallet app
fun triggerDeepLink() {
    val currentSessionState = dApp.getInitialSessionState() ?: return
    try {
        val myIntent = Intent(Intent.ACTION_VIEW, Uri.parse(currentSessionState.connectionParams.toUri()))
        startActivity(myIntent)
    } catch (_: ActivityNotFoundException) {
        Toast
                .makeText(requireContext(),
                          "No application can handle this request. Please install a wallet app",
                          Toast.LENGTH_LONG)
                .show()
    }
}
```

</details>

<details>
<summary>Wallet</summary>


**Init**

```kotlin
val connectionParams: ConnectionParams // get through deeplink, QR code ...

val initialSessionState = InitialSessionState(
        connectionParams,
        myPeerId = UUID.randomUUID().toString(),
        myPeerMeta = PeerMeta(
                name = "Wallet",
                url = "https://wallet.com",
                description = "Wallet Description",
                icons = listOf("https://img.favpng.com/1/20/24/wallet-icon-png-favpng-TQrAD3mHXn7Yey6wnt6aa97YF.jpg")
        )
)

val wallet: Wallet = createWallet(sessionStoreName = "...")
```

**Open Socket**

```kotlin
wallet.openSocketAsync(initialSessionState,
                       callback = ::onSessionCallback,
                       onOpen = { freshOpened ->
                           // sendSessionRequest() etc...
                       })
// or
coroutineScope.launch(dispatcherProvider.io()) {
    val freshOpened = wallet.openSocket(initialSessionState,
                                        callback = ::onSessionCallback)
    // sendSessionRequest() etc...
}
```

**Close**

```kotlin
wallet.closeAsync(deleteLocal = false,
                  deleteRemote = false,
                  onClosed = { freshClosed ->

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.7.0] - 2022-09-14

### Added

- [WalletConnectCore.onClosing](walletconnect/src/main/java/walletconnect/WalletConnectCore.kt) protected method, called
  in `closeAsync` so `DAppManager`/`WalletManager` can do cleanup
- [DApp.sendRequest](walletconnect-core/src/main/java/walletconnect/core/DApp.kt) has non-suspending
  overload `sendRequestAsync` with `onRequested`, `onRequestError`, `onCallback` parameters
- [SessionLifecycle.openSocketAsync](walletconnect-core/src/main/java/walletconnect/core/session/SessionLifecycle.kt)
  has suspending `openSocket` overload
- [SessionLifecycle.closeAsync](walletconnect-core/src/main/java/walletconnect/core/session/SessionLifecycle.kt)
  has suspending `close` overload
- [SessionLifecycle.disconnectSocketAsync](walletconnect-core/src/main/java/walletconnect/core/session/SessionLifecycle.kt)
  has suspending `disconnectSocket` overload
- [SessionLifecycle.reconnectSocketAsync](walletconnect-core/src/main/java/walletconnect/core/session/SessionLifecycle.kt)
  has suspending `reconnectSocket` overload

### Changed

- [WalletConnectCore.deleteSessionInternal](walletconnect/src/main/java/walletconnect/WalletConnectCore.kt) method
  is `private` now
- [CallbackData.simplifiedName](walletconnect-core/src/main/java/walletconnect/core/session/callback/CallbackData.kt)
  has `withContent:Boolean` parameter to get detailed message.
- Overridden `FailureType.toString` & `SignType.toString` methods
- [Wallet.approveRequest](walletconnect-core/src/main/java/walletconnect/core/Wallet.kt)
  , [Wallet.rejectRequest](walletconnect-core/src/main/java/walletconnect/core/Wallet.kt) are not `suspend` function
  anymore. Wallet do not need `messageId` (already knows from params). If it needs callback, callback is already called
  with `messageId`
- Refactor sample code & README with new overloaded methods

## [0.6.3] - 2022-08-22

### Added

- [walletconnect-requests](walletconnect-requests) module
  , [SwitchEthChain](walletconnect-requests/src/main/java/walletconnect/requests/wallet/SwitchEthChain.kt)
  , [CustomRpcMethods](walletconnect-requests/src/main/java/walletconnect/requests/CustomRpcMethods.kt)
- `SwitchEthChain` button in sample app
- Proguard rule in Readme for `walletconnect-requests` module

### Changed

- [Wallet.approveRequest](walletconnect-core/src/main/java/walletconnect/core/Wallet.kt) `result:Any` -> `result:Any?`

## [0.6.2] - 2022-08-16

### Changed

- [SessionStore#getAll()](walletconnect-core/src/main/java/walletconnect/core/session_state/SessionStore.kt)
  returns `Set` instead of `List`, like `getAllAsFlow()`

## [0.6.1] - 2022-08-16

### Added

- [Cryptography#generateSymmetricKey()](walletconnect-core/src/main/java/walletconnect/core/cryptography/Cryptography.kt)

### Changed

- Readme

## [0.6.0] - 2022-08-08

### Changed

- [DAppManager](walletconnect/src/main/java/walletconnect/DAppManager.kt)
  , [WalletManager](walletconnect/src/main/java/walletconnect/WalletManager.kt)
  , [WalletConnectCore](walletconnect/src/main/java/walletconnect/WalletConnectCore.kt) constructors takes `Socket` as
  parameter instead of `socketFactory`

### Added

- Dokka plugin for kotlin Javadoc (still may not work)

### Fixed

- `walletconnect-adapter-gson` and `walletconnect-adapter-moshi` modules artifacts bug

## [0.5.0] - 2022-08-08

### Added

- Working Android sample for dApp, wallet communication
- Serialization/Deserialization adapter using [Gson](https://github.com/google/gson)
  / [Moshi](https://github.com/square/moshi)
- Store/Restore Sessions using File/SharedPreferences
- Socket implementation using [Scarlet](https://github.com/tinder/scarlet)
- Pure kotlin modules
- Unit tests (except `:walletconnect` module)

[Unreleased]: https://github.com/jemshit/walletconnect/compare/main...develop

[0.7.0]: https://github.com/jemshit/walletconnect/compare/0.6.3..0.7.0

[0.6.3]: https://github.com/jemshit/walletconnect/compare/0.6.2..0.6.3

[0.6.2]: https://github.com/jemshit/walletconnect/compare/0.6.1..0.6.2

[0.6.1]: https://github.com/jemshit/walletconnect/compare/0.6.0..0.6.1

[0.6.0]: https://github.com/jemshit/walletconnect/compare/0.5.0..0.6.0

[0.5.0]: https://github.com/jemshit/walletconnect/releases/tag/0.5.0

sample/src/main/java/walletconnect/sample/DAppFragment.kt
/**
 * MIT License
 * Copyright (c) 2022 Jemshit Iskenderov
 */
package walletconnect.sample

import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import android.widget.Toast.LENGTH_LONG
import android.widget.Toast.LENGTH_SHORT
import androidx.core.content.ContextCompat
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.coroutineScope
import androidx.lifecycle.lifecycleScope
import com.google.android.material.bottomsheet.BottomSheetDialog
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import walletconnect.core.DApp
import walletconnect.core.cryptography.toHex
import walletconnect.core.requests.eth.EthTransaction
import walletconnect.core.requests.eth.SignType.Companion.toMethod
import walletconnect.core.session.callback.*
import walletconnect.core.session.model.InitialSessionState
import walletconnect.core.session.model.json_rpc.EthRpcMethod
import walletconnect.core.session_state.model.PeerMeta
import walletconnect.requests.CustomRpcMethods
import walletconnect.requests.wallet.SwitchChain
import walletconnect.sample.bnb_provider.BnbViewModel
import walletconnect.sample.databinding.DappFragmentBinding
import walletconnect.sample.sign.SignFragment
import walletconnect.sample.store.SessionListFragment
import java.math.BigInteger
import java.util.*

class DAppFragment : BaseFragment() {

    private var _binding: DappFragmentBinding? = null
    private val binding get() = _binding!!

    override val fragmentTag: String = javaClass.simpleName
    private lateinit var dApp: DApp
    override val initialSessionState by lazy {
        InitialSessionState(
                connectionParams,
                myPeerId = UUID.randomUUID().toString(),
                myPeerMeta = PeerMeta(
                        name = "DApp",
                        url = "https://dapp.com",
                        description = "DApp Description",
                        icons = listOf("https://www.dapp.com/img/Icon_Logotype_2.png")
                )
        )
    }
    private val bnbVM: BnbViewModel by activityViewModels()

    // region lifecycle
    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?)
            : View {
        super.onCreateView(inflater, container, savedInstanceState)

        _binding = DappFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View,
                               savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        dApp = createDApp(sessionStoreName = fragmentTag)
        sessionLifecycle = dApp

        setClickActions()
        observeSignRequestEvents()
    }

    override fun consoleTextView(): TextView = binding.textConsole

    override fun onConsoleUpdate() {
        binding.scrollView.post {
            // then wait for the child of the scroll view to be laid out
            binding.scrollView.getChildAt(0).postDelayed(
                    {
                        // finally scroll
                        binding.scrollView.smoothScrollBy(0, binding.scrollView.bottom)
                    },
                    500L
            )
        }
    }

    override fun onDestroyView() {
        // [super.onDestroyView] calls [dApp.close], which expects caller to trigger [SessionCallback]
        onSessionCallback(SessionCallback.SessionClosedLocally)
        super.onDestroyView()
        _binding = null
    }
    // endregion

    private fun setClickActions() {
        binding.buttonClear.setOnClickListener {
            binding.textConsole.text = ""
        }
        binding.buttonOpen.setOnClickListener {
            dApp.openSocketAsync(initialSessionState,
                                 callback = ::onSessionCallback,
                                 onOpened = null)
        }
        binding.buttonClose.setOnClickListener {
            dApp.closeAsync(deleteLocal = false,
                            deleteRemote = false)
        }
        binding.buttonDelete.setOnClickListener {
            dApp.closeAsync(deleteLocal = true,
                            deleteRemote = true)
        }

        binding.buttonSendSessionRequest.setOnClickListener {
            dApp.sendSessionRequest(null)
        }
        binding.buttonSwitchChain.setOnClickListener {
            if (!approvedAddress.isNullOrBlank()) {
                dApp.sendRequestAsync(CustomRpcMethods.SwitchEthChain,
                                      data = listOf(SwitchChain("0x" + 56.toHex())),
                                      itemType = SwitchChain::class.java)
            }
        }

        binding.buttonSendSignRequest.setOnClickListener {
            if (!approvedAddress.isNullOrBlank()) {
                SignFragment
                        .create(messageId = null, approvedAddress!!, message = null)
                        .show(requireActivity().supportFragmentManager, fragmentTag)
            }
        }
        binding.buttonSignTx.setOnClickListener {
            if (!approvedAddress.isNullOrBlank()) {
                dApp.sendRequestAsync(EthRpcMethod.SignTransaction,
                                      data = listOf(createTransaction()),
                                      itemType = EthTransaction::class.java)
            }
        }
        binding.buttonSendTx.setOnClickListener {
            if (!approvedAddress.isNullOrBlank()) {
                dApp.sendRequestAsync(EthRpcMethod.SendTransaction,
                                      data = listOf(createTransaction()),
                                      itemType = EthTransaction::class.java)
            }
        }
        binding.buttonSendTokenTx.setOnClickListener {
            if (!approvedAddress.isNullOrBlank()) {
                if (bnbVM.gas.value.isBlank()) {
                    bnbVM.estimateGas(createTokenTransaction("", ""),
                                      isTestNet = approvedChainId!! == 97)
                    Toast.makeText(requireContext(), "Fetching GAS. Try Again", LENGTH_SHORT).show()

                } else if (bnbVM.gasPrice.value.isBlank()) {
                    bnbVM.getGasPrice(isTestNet = approvedChainId!! == 97)
                    Toast.makeText(requireContext(), "Fetching GAS PRICE. Try Again", LENGTH_SHORT).show()

                } else {
                    dApp.sendRequestAsync(EthRpcMethod.SendTransaction,
                                          data = listOf(createTokenTransaction(bnbVM.gas.value,
                                                                               bnbVM.gasPrice.value)),
                                          itemType = EthTransaction::class.java)
                }
            }
        }
        binding.buttonDeepLink.setOnClickListener {
            triggerDeepLink()
        }

        binding.imageSessionList.setOnClickListener {
            SessionListFragment
                    .create(fragmentTag)
                    .show(requireActivity().supportFragmentManager, fragmentTag)
        }
    }

    private fun triggerDeepLink() {
        val currentSessionState = dApp.getInitialSessionState() ?: return
        try {
            val myIntent = Intent(Intent.ACTION_VIEW, Uri.parse(currentSessionState.connectionParams.toUri()))
            startActivity(myIntent)
        } catch (_: ActivityNotFoundException) {
            Toast
                    .makeText(requireContext(),
                              "No application can handle this request. Please install a wallet app",
                              Toast.LENGTH_LONG)
                    .show()
        }
    }

    override fun onSessionCallback(callbackData: CallbackData) {
        viewLifecycleOwner.lifecycle.coroutineScope.launch(dispatcherProvider.ui()) {
            val stateText = updateLatestState(callbackData)
            if (stateText != null) {
                binding.textCallbacks.text = stateText
            }

            when (callbackData) {
                is SessionCallback -> {
                    when (callbackData) {
                        is SessionCallback.LocalSessionStateUpdated -> {
                            if (callbackData.sessionState == null) {
                                binding.imageMyLogo.setImageDrawable(null)
                                binding.textMyId.text = "Me(null)"
                                binding.imageRemoteLogo.setImageDrawable(null)
                                binding.textRemoteId.text = "Remote(null)"
                                binding.textChainId.text = "null"
                                binding.textAccounts.text = "null"
                                approvedAddress = null
                                approvedChainId = null
                            } else {
                                binding.imageMyLogo.loadSvgImage(callbackData.sessionState!!.myPeerMeta.icons?.firstOrNull())
                                binding.textMyId.text = "Me(${callbackData.sessionState!!.myPeerId.take(6)})"
                                binding.imageRemoteLogo.loadSvgImage(callbackData.sessionState!!.remotePeerMeta.icons?.firstOrNull())
                                binding.textRemoteId.text = "Rem.(${callbackData.sessionState!!.remotePeerId.take(6)})"
                                binding.textChainId.text = "Chain:${callbackData.sessionState!!.chainId?.toString()}"
                                binding.textAccounts.text =
                                    "Acc:${callbackData.sessionState!!.accounts?.joinToString { it.take(6) }}"
                                approvedAddress = callbackData.sessionState!!.accounts?.firstOrNull()
                                approvedChainId = callbackData.sessionState!!.chainId
                            }
                        }
                        else -> {}
                    }
                }
                is SocketCallback -> {
                    when (callbackData) {
                        SocketCallback.SocketConnecting -> {
                            binding.textTitle.setTextColor(ContextCompat.getColor(requireContext(), R.color.orange))
                        }
                        SocketCallback.SocketClosed -> {
                            binding.textTitle.setTextColor(ContextCompat.getColor(requireContext(), R.color.white))
                        }
                        SocketCallback.SocketConnected -> {
                            binding.textTitle.setTextColor(ContextCompat.getColor(requireContext(), R.color.green))
                        }
                        SocketCallback.SocketDisconnected -> {
                            binding.textTitle.setTextColor(ContextCompat.getColor(requireContext(), R.color.red))
                        }
                        is SocketCallback.SocketMessage -> {

                        }
                    }
                }
                is RequestCallback -> {
                    when (callbackData) {
                        is RequestCallback.EthSignRequested -> {
                            //triggerDeepLink()
                        }
                        is RequestCallback.EthSignTypedDataRequested -> {
                            //triggerDeepLink()
                        }
                        is RequestCallback.EthSignResponse -> {
                            BottomSheetDialog(requireContext())
                                    .apply {
                                        setContentView(R.layout.text_item)
                                        val textView = findViewById<TextView>(R.id.textContent)
                                        textView?.text = "Signature:\n" + callbackData.signature
                                    }
                                    .show()
                        }

                        is RequestCallback.CustomRequested -> {

                        }
                        is RequestCallback.CustomResponse -> {

                        }

                        is RequestCallback.EthSignTxRequested -> {

                        }
                        is RequestCallback.EthSignTxResponse -> {

                        }
                        is RequestCallback.EthSendRawTxRequested -> {

                        }
                        is RequestCallback.EthSendRawTxResponse -> {

                        }
                        is RequestCallback.EthSendTxRequested -> {

                        }
                        is RequestCallback.EthSendTxResponse -> {

                        }

                        is RequestCallback.RequestRejected -> {
                            BottomSheetDialog(requireContext())
                                    .apply {
                                        setContentView(R.layout.text_item)
                                        val textView = findViewById<TextView>(R.id.textContent)
                                        textView?.text = "Rejected:\n" + callbackData.error.message
                                    }
                                    .show()
                        }
                    }
                }
                is FailureCallback -> {

                }
            }
        }
    }

    // region Sign/Transaction
    private fun observeSignRequestEvents() {
        viewModel.signRequests
                .onEach { ethSign ->
                    try {
                        ethSign.validate()
                        dApp.sendRequest(
                                method = ethSign.type.toMethod(),
                                data = ethSign.toList(),
                                itemType = String::class.java
                        )
                    } catch (error: Exception) {
                        Toast.makeText(requireContext(), error.message, LENGTH_LONG).show()
                    }
                }
                .launchIn(viewLifecycleOwner.lifecycleScope)
    }

    private fun createTransaction()
            : EthTransaction {
        return EthTransaction(
                from = approvedAddress!!,
                to = "0x621261D26847B423Df639848Fb53530025a008e8",
                data = "",
                chainId = if (approvedChainId == null) null else "0x" + approvedChainId!!.toHex(),

                gas = null,
                gasPrice = null,
                gasLimit = null,
                maxFeePerGas = null,
                maxPriorityFeePerGas = null,

                value = "0x" + BigInteger("10000000000000000").toString(16), // 1_000_000_000_000_000_000L.toHex(),
                nonce = null
        )
    }

    private fun createTokenTransaction(gas: String,
                                       gasPrice: String)
            : EthTransaction {
        // bnb=56, bnb_test=97
        val chainId = "0x" + approvedChainId!!.toHex()
        // on bnb_test
        val smartContractAddressMain = "0xb465f3cb6Aba6eE375E12918387DE1eaC2301B05"
        val smartContractAddress = "0xcD0FC5078Befa4701C3692F4268641A468DEB8d5"
        // keccak256(transactionMethod).take(4 bytes)
        val smartContractTransactionMethod = "a9059cbb"
        // must be 32 bytes
        val to = "621261D26847B423Df639848Fb53530025a008e8".padStart(64, '0')
        // 3 decimal; must be 32 bytes; 1_000L.toHex()
        val amount = BigInteger("1000").toString(16).padStart(64, '0')

        return EthTransaction(
                from = approvedAddress!!,
                to = if (approvedChainId!! == 56) smartContractAddressMain else smartContractAddress,
                data = "0x$smartContractTransactionMethod$to$amount",
                chainId = chainId,

                gas = gas, // can't auto calculate for smart contract
                gasPrice = gasPrice, // can auto calculate
                gasLimit = null,
                maxFeePerGas = null,
                maxPriorityFeePerGas = null,

                value = "", // this is in 'data' field now
                nonce = null // ignored by metamask
        )
    }
    // endregion
}